File: ios/telobike/telobike/TBObserver.m
//
//  TBObserver.m
//
//  Created by Elad Ben-Israel on 4/24/14.
//  Copyright (c) 2014 Citylifeapps. All rights reserved.
//

#import "TBObserver.h"

@interface TBObserver ()

@property (strong, nonatomic) id object;
@property (strong, nonatomic) NSString *keyPath;
@property (strong, nonatomic) void(^block)(void);

@end

@implementation TBObserver

- (instancetype)initWithObject:(id)object keyPath:(NSString *)keyPath block:(void(^)(void))block
{
    if (!object || keyPath.length == 0) {
        return nil;
    }
    
    self = [super init];
    if (self) {
        self.object = object;
        self.keyPath = keyPath;
        self.block = block;
        [self.object addObserver:self forKeyPath:self.keyPath options:NSKeyValueObservingOptionInitial context:nil];
    }
    return self;
}

+ (instancetype)observerForObject:(id)object keyPath:(NSString *)keyPath block:(void(^)(void))block
{
    return [[TBObserver alloc] initWithObject:object keyPath:keyPath block:block];
}

- (void)dealloc
{
    [self.object removeObserver:self forKeyPath:self.keyPath];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (self.block) {
        dispatch_async(dispatch_get_main_queue(), self.block);
    }
}

@end

File: ios/telobike/telobike/TBObserver.swift
//
//  TBObserver.swift
//  telobike
//
//  Created by Elad Ben-Israel on 9/19/14.
//  Copyright (c) 2014 Elad Ben-Israel. All rights reserved.
//

import Foundation

@objc class TBObserver: NSObject {
    private let object: NSObject
    private let keyPath: String
    private let block: () -> ()
    
    private init(object: NSObject, keyPath: String, block: () -> ()) {
        self.object = object
        self.keyPath = keyPath
        self.block = block
        super.init()
        self.object.addObserver(self, forKeyPath: self.keyPath, options: .Initial, context: nil)
    }
    
    deinit {
        self.object.removeObserver(self, forKeyPath: self.keyPath)
    }
    
    override func observeValueForKeyPath(keyPath: String, ofObject object: AnyObject, change: [NSObject : AnyObject], context: UnsafeMutablePointer<Void>) {
        dispatch_async(dispatch_get_main_queue(), self.block)
    }
    
    class func observerForObject(object: NSObject?, keyPath: String?, block: () -> ()) -> TBObserver? {
        if object == nil || keyPath == nil || keyPath!.isEmpty {
            return nil
        }
        return TBObserver(object: object!, keyPath: keyPath!, block: block)
    }
}